### ISSUE TITLE
Create a splash damage optimiser

### ISSUE THREAD

**Issue by Unusual229** [RaphaelRoyerRivard/MicroMachine#1095]
Based on this info:
![image](https://user-images.githubusercontent.com/8284324/97257897-933f0180-17ed-11eb-9099-64d3e97ee411.png)

The optimiser needs to be able to handle:
Tanks, widow mines, nukes, PF, Archon, purification nova (disruptor), Thor (air?), Liberator air and banelings. Storms and such spells could be useful as well.

We should handle other unit types than just Terran's so we can #1 change race one day maybe, #2 consider the opponents best move (like a burrowed widow mine, or an HT, etc). Shouldn't be more work anyway (except for spells maybe).

The function should these parameters :
-weapon/Unittype/spell ID, to know the damage of the attack
-list of pairs <float splash range, float weapon damage %> (from the image above), a range of -1 should mean dmg to the main target only (so none splash), applies on to widow mines though.
-(CCPosition) The attacking/casting unit position
-(float) The range of the attack/cast around the unit position (unrelated to the splash range)
-(bool) ignore buildings
-(bool) ignore units (for nukes)
-(bool) if the target has to be a unit or not. Tanks, for example, have to, while HT storms don't.

Later on, this should be expanded to handle :
-Hellion, Colossus, both attack in a straight line, could be the same code.
-Hellbat, attacks in a cone, unique attack style.
-Ultralisk, half circle, might be possible to adapt it in the circle splash behavior.
-Lurker, relatively similar to Hellion and Colossus, but is a line of circles.
-Mutalisk, cannot predict the exact targets of the bounces, since its random, but could calculate to total maximum damage if we consider the bounce maximum range

**Comment by Unusual229** [comment]
Could maybe also make it work to calculate the maximum number of units that can be hit by a none-damaging spell :
Ghost (EMP Round): AoE spell
Raven (Anti-Armor Missile): AoE spell
Infestor (Fungal Growth): AoE spell

**Comment by Unusual229** [comment]
The way I would consider the damage done and optimise it would be :
for each enemy unit hit
total += ((Unit value * damage done in %)  + Unit value if unit dies)

for each ally unit hit
total -= 1.5x ((Unit value * damage done in %) + 2x Unit value if unit dies)

Adding the unit value if it dies is to prioritize kills to some extent, but not over doing it.
I'd consider the splash on our units as quite negative, I'd rather not have our splash units do more or similar damage to us than to our opponent. I'd much rather not lose units to splash, since we can heal them, however for a valuable shot it can be acceptable (losing a marine to kill some roaches or something).

**Comment by RaphaelRoyerRivard** [comment]
For the parameters, the first and third one should be replaced by the unit, because its damage can vary depending on its target. The method Util::GetDamageForTarget(attacker, target) will be useful. The method Util::GetAttackRangeForTarget(attacker, target) will also be useful so we don't need to have a range parameter. The rest looks good to me.

**Comment by Unusual229** [comment]
Could have a method overload that allows to pass an array of allied units to check and an array of enemy units to check (used to specify close by threat), but also allow us to use all units by default.

**Comment by Unusual229** [comment]
Depending how getting close threats is programmed, we might need to use the "all units" version of the method to handle tanks and such long range  units.

**Comment by RaphaelRoyerRivard** [comment]
You can pull my branch to get the empty method `MicroManager::getPriorityOfTargetConsideringSplash(attacker, target)` that gets called when the unit has a splash attack.

**Comment by Unusual229** [comment]
So you will call this method for each target you can hit and pick the best one?

**Comment by RaphaelRoyerRivard** [comment]
Yes, it's called in the function that returns a float which is a target
priority value. All the checks to see if the target is one we might not
want to shoot, like egg, larva, cloaked, burrowed, etc. are already passed
so the method is called only for good targets. But those checks should
probably be done also in that method for other units that might receive
splash damage.

Raphaël Royer-Rivard

**Comment by Unusual229** [comment]
Splash will be checked on all units, so it would included workers and such if they are next to marines or tank or whatever.
